--- ./build/source-files/libgc-6.8/os_dep.c-orig	2011-05-04 10:03:26.000000000 -0700
+++ ./build/source-files/libgc-6.8/os_dep.c	2011-05-04 10:03:53.000000000 -0700
@@ -977,7 +977,7 @@
       c = stat_buf[buf_offset++];
     }
     close(f);
-    if (result < 0x10000000) ABORT("Absurd stack bottom value");
+    if (result < 0x1000) ABORT("Absurd stack bottom value");
     return (ptr_t)result;
   }
 
