--- build-virgin/source-files/vnc4-4.1.1+X4.3.0/common/rfb/ZoogCanvasController.h	1969-12-31 16:00:00.000000000 -0800
+++ build/source-files/vnc4-4.1.1+X4.3.0/common/rfb/ZoogCanvasController.h	2012-03-11 09:31:03.000000000 -0700
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <set>
+#include <rfb/SConnection.h>
+#include <rfb/SMsgWriter.h>
+#include <rfb/VNCServerST.h>
+#include <rfb/DisplayIfc.h>
+#include "pal_abi/pal_abi.h"
+
+namespace rfb {
+
+	class VNCZoogDisplay;
+
+	class ZoogCanvasController :
+		public SocketIfc
+	{
+	public:
+		ZoogCanvasController(VNCZoogDisplay *_zoog_display);
+
+		// SocketIfc
+		virtual network::Socket* getSock();
+		virtual void processMessages();
+
+	private:
+		int listensock;
+		VNCZoogDisplay *zoog_display;
+		network::Socket* sock;
+	};
+}
